contracts/base/Dispatcher.sol

```diff
@@ -33,171 +37,207 @@ abstract contract Dispatcher is Payments, V2SwapRouter, V3SwapRouter, Callbacks,
-                    IAllowanceTransfer.PermitSingle calldata permitSingle;
-                    assembly {
-                        permitSingle := inputs.offset
+                    if (selector != IERC721Permit.permit.selector) {
+                        revert InvalidAction(selector);
                     }
-                    bytes calldata data = inputs.toBytes(6); // PermitSingle takes first 6 slots (0..5)
-                    PERMIT2.permit(lockedBy, permitSingle, data);
-                } else if (command == Commands.WRAP_ETH) {
-                    // equivalent: abi.decode(inputs, (address, uint256))
-                    address recipient;
-                    uint256 amountMin;
+
+                    (success, output) = address(V3_POSITION_MANAGER).call(inputs);
+                } else if (command == Commands.V3_POSITION_MANAGER_CALL) {
+                    bytes4 selector;
+                    uint256 tokenId;
                     assembly {
-                        recipient := calldataload(inputs.offset)
-                        amountMin := calldataload(add(inputs.offset, 0x20))
+                        selector := calldataload(inputs.offset)
+                        // tokenId is always the first parameter in the valid actions
+                        tokenId := calldataload(add(inputs.offset, 0x04))
                     }
-                    Payments.wrapETH(map(recipient), amountMin);
-                } else if (command == Commands.UNWRAP_WETH) {
-                    // equivalent: abi.decode(inputs, (address, uint256))
-                    address recipient;
-                    uint256 amountMin;
-                    assembly {
-                        recipient := calldataload(inputs.offset)
-                        amountMin := calldataload(add(inputs.offset, 0x20))
+
+                    if (!isValidAction(selector)) {
+                        revert InvalidAction(selector);
                     }
-                    Payments.unwrapWETH9(map(recipient), amountMin);
-                } else if (command == Commands.PERMIT2_TRANSFER_FROM_BATCH) {
-                    (IAllowanceTransfer.AllowanceTransferDetails[] memory batchDetails) =
-                        abi.decode(inputs, (IAllowanceTransfer.AllowanceTransferDetails[]));
-                    permit2TransferFrom(batchDetails, lockedBy);
-                } else if (command == Commands.BALANCE_CHECK_ERC20) {
-                    // equivalent: abi.decode(inputs, (address, address, uint256))
-                    address owner;
-                    address token;
-                    uint256 minBalance;
-                    assembly {
-                        owner := calldataload(inputs.offset)
-                        token := calldataload(add(inputs.offset, 0x20))
-                        minBalance := calldataload(add(inputs.offset, 0x40))
+                    if (!isAuthorizedForToken(lockedBy, tokenId)) {
```

Comment:
This is a little bit of an odd authorization scheme.

This function call checks `isAuthorizedForToken` which checks if the msg.sender is the owner, or the approved "operator" (the address that gets set in `Position` struct) or if the operator mapping in base ERC721 contract is set to true.

By using this authorization logic inside the migrator we essentially require 2 operators to already be set on the token. (One, this contract must be approved because its going to be doing the call on position manager, but now 2, the current msg.sender needs to be approved bc we read it again here.) OR we just force the msg.sender is always the owner) 

the other weird thing is that only one of those "approvals" is permittable (setting the operator address in the Position struct is the only way to do a signature based permit on the ERC721). This is not necessarily a problem with this design here except for the fact that maybe there is a case to not use the authorization scheme of the position manager contract, and instead have our own authorization mapping inside the router. Food for thought. FWIW I also think this is a shortcoming of the position manager

Reply:
we discussed on slack, going to document well